src/mlpack/methods/ann/layer/linear3d_impl.hpp

```diff
@@ -0,0 +1,150 @@
+/**
+ * @file methods/ann/layer/linear3d_impl.hpp
+ * @author Mrityunjay Tripathi
+ *
+ * Implementation of the Linear layer class which accepts 3D input.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_LINEAR3D_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_LINEAR3D_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "linear3d.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType,
+    typename RegularizerType>
+Linear3D<InputDataType, OutputDataType, RegularizerType>::Linear3D() :
+    inSize(0),
+    outSize(0)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType,
+    typename RegularizerType>
+Linear3D<InputDataType, OutputDataType, RegularizerType>::Linear3D(
+    const size_t inSize,
+    const size_t outSize,
+    RegularizerType regularizer) :
+    inSize(inSize),
+    outSize(outSize),
+    regularizer(regularizer)
+{
+  weights.set_size(outSize * inSize + outSize, 1);
+}
+
+template<typename InputDataType, typename OutputDataType,
+    typename RegularizerType>
+void Linear3D<InputDataType, OutputDataType, RegularizerType>::Reset()
+{
+  weight = arma::mat(weights.memptr(), outSize, inSize, false, false);
+  bias = arma::mat(weights.memptr() + weight.n_elem,
+      outSize, 1, false, false);
+}
+
+template<typename InputDataType, typename OutputDataType,
+    typename RegularizerType>
+template<typename eT>
+void Linear3D<InputDataType, OutputDataType, RegularizerType>::Forward(
+    const arma::Mat<eT>& input, arma::Mat<eT>& output)
+{
+  Log::Assert(input.n_rows % inSize == 0);
+
+  const size_t nPoints = input.n_rows / inSize;
+  const size_t batchSize = input.n_cols;
+
+  output.set_size(outSize * nPoints, batchSize);
+
+  arma::Cube<eT> inputTemp(const_cast<arma::Mat<eT>&>(input).memptr(), inSize,
+      nPoints, batchSize, true, false);
+
+  for (size_t i = 0; i < batchSize; ++i)
+  {
+    output.col(i) = arma::vectorise(weight * inputTemp.slice(i)
+        + arma::repmat(bias, 1, nPoints));
+  }
+}
+
+template<typename InputDataType, typename OutputDataType,
+    typename RegularizerType>
+template<typename eT>
+void Linear3D<InputDataType, OutputDataType, RegularizerType>::Backward(
+    const arma::Mat<eT>& /* input */,
+    const arma::Mat<eT>& gy,
+    arma::Mat<eT>& g)
+{
+  Log::Assert(gy.n_rows % inSize == 0);
+
+  const size_t nPoints = gy.n_rows / outSize;
```

Comment:
I think there should be either `inSize` or `outSize` in both cases.

Reply:
Ohh... yes. It should be `outSize`.